lsql/judge/views.py

```diff
@@ -354,6 +354,30 @@ def show_achievements(request, user_id):
                                                  'username': this_user.username})
 
 
+@login_required
+def show_hints(request):
+    """View for show the used hints"""
+    this_user = get_user_model().objects.get(pk=request.user.pk)
+    context = {'user': this_user.username, 'elements': []}
+    elements = []
+    problems = []
+    hints = UsedHint.objects.filter(user=request.user.pk).order_by('request_date')
+    for hint in hints:
+        if hint.hint_definition.problem.title_html in problems:
```

Comment:
No es eficiente insertar en una lista para luego buscar dentro de ella, para eso tienes los diccionarios en Python.

En este caso, lo mejor es que elements fuera un diccionario así: 

   {'titulo del problema': [lista de pistas]}

Así te quitas la necesidad de la lista problems, ya que podrías consultar directamente elements para saber si un problema está o no

Reply:
Y para la plantilla no es problema, porque puedes recorrer diccionario.items dentro de la plantilla y te devuelve la clave (título de problema) y las pistas (lista de UsedHint)

Reply:
> No es eficiente insertar en una lista para luego buscar dentro de ella, para eso tienes los diccionarios en Python.
> En este caso, lo mejor es que elements fuera un diccionario así:
> ```
> {'titulo del problema': [lista de UsedHint]}
> ```
> Así te quitas la necesidad de la lista problems, ya que podrías consultar directamente elements para saber si un problema está o no

Al principio iba a hacerlo asi, pero para cada pista usada tendría que recorrer cada elemento de la lista de diccionarios y veía menos lio el tener una lista solo con los problemas usados. Que para acceder al nombre de problema no lo puedo comprobar como antes con un: `if hint.hint_definition.problem.title_html in elements:` tendría que mirar cada diccionario dentro de la lista de `elements`

Reply:
No entiendo del todo esa necesidad. Para la plantilla lo que necesitas es tomar todas las pistas usadas de un usuario y agruparlas por problema. Si usas un diccionario `{'titulo del problema': [lista de UsedHint]}` y recorres un `QuerySet` de `UsedHint`, cuando examines un `UsedHint` lo metes en la lista que le toque (o creas una entrada al dict). Pero no tienes que buscar nada en ninguna lista porque accedes usando la clave del diccionario. 

> tendría que mirar cada diccionario dentro de la lista de elements

Es que `elements` no debe ser una lista de diccionarios sino un diccionario de listas: accedes rápido usando la clave y añades la pista usada al final de la lista actual.

Una cosa que es bastante fea es usar el título del problema (que nadie asegura que sea único) como clave para agrupar pistas usadas. Realmente lo que deberíamos hacer es agrupar por PK de problema (que es numérica y es única), y para saber el título del problema tirar de las claves foráneas UsedHint -> Hint -> Problem

    {<PK del problema>: [lista de UsedHint]}

Reply:
Vale lo tengo pero con la clave siendo el nombre del problema, si pongo el PK del problema no lo tengo muy claro:

```
{% for key, values in elements.items %}
<h2>{% translate "Problema: " %}<span class="headAchievement"><a href="{% url 'judge:problem' key %}">
    {{ values.hint_definition.problem.title_html|safe }}</a></span></h2>
```